packages/core/src/types/value.ts

```diff
@@ -19,7 +20,7 @@ export interface ShapeVal<T> {
   contents: RealShape<T>;
 }
 
-export type ArgValWithSourceLoc<T> = ArgVal<T> & SourceRange;
+export type ArgValWithExpr<T> = ArgVal<T> & { expr: ResolvedExpr<A> };
```

Comment:
Generally I didn't want to use `C` (for no particular reasons). I used `A` but all the location information are placed in using the `...` operator.

Reply:
I understand the data will actually be inserted to the nodes, but I'd prefer using `C` as much as possible so we can avoid using `isConcrete` for better type safety. `isConcrete` is kind of an escape hatch iirc. I know we haven't been doing this properly in, say, error reporting, but would love to move towards better-typed errors in general. Any chance to improve this situation?

If `A` and `C` are just meaningless and we end up mixing synthetic and concrete nodes all the time, I'm starting to question the need for them in the first place...

Reply:
The reason why many expressions are still abstract nodes is that many expressions are generated. An example of this is sampled shape properties.

Here, `ArgValWithExpr` is used in function (objective, constraint) calls. The problem is the constraint `onCanvas` which can automatically be inserted, and hence has no location information on the arguments.

Previously we just made a fake AST `Path` node with some arbitrary location information

Reply:
In particular, everything before `translate` we should be using `C` since we are only accessing the actual AST. Since `translate` can introduce things that have no location information, everything after `translate` we use `A`.